### PR TITLE
Added 3rd endpoint

### PR DESCRIPTION
**Pull request by yiancar**
- Added 3rd endpoint check in asmcommon. This adds 2 extra cycles to the USB routine when enabled. With 16MHz clock (most if not all qmk build use that) the routine is still within spec.

- Updated prototype.h file to add additional options.

- Simulatanius changes in QMK: https://github.com/qmk/qmk_firmware/pull/9020

### CODE REVIEW

**Comment by fauxpark** [comment]
I'm pretty sure all of the PS2AVRGB boards (the majority of V-USB boards) run at 12MHz.

**Comment by yiancar** [comment]
Interesting, Sorry I was unaware. Could you try with some if you have any and report results. Maybe A good way is to add a warning if 3 endpoints are used at the same time. In theory disabling the 3rd endpoint should take you back to the way everything was before.

**Comment by yiancar** [comment]
Ok Back for more:
This has now been tested at 12MHz and 16MHz and works as expected.
Complimentary PR at future branch can verify this.

File: usbdrv/usbdrvasm.S [hunk 1]
```diff
@@ -293,7 +295,7 @@ usbCrc16Append:
 #   define cnt16H   r31
 #   define cntH     r18
 
-#else  /* __IAR_SYSTEMS_ASM__ */ 
+#else  /* __IAR_SYSTEMS_ASM__ */
```
Comment (fauxpark) [inline usbdrv/usbdrvasm.S]:
```suggestion
#else  /* __IAR_SYSTEMS_ASM__ */ 
```

File: usbdrv/usbdrvasm.S [hunk 2]
```diff
@@ -161,7 +163,7 @@ RTMODEL "__rt_version", "3"
 ; unsigned usbCrc16(unsigned char *argPtr, unsigned char argLen)
 ; {
 ; unsigned crc = 0xffff;
-; 
+;
```
Comment (fauxpark) [inline usbdrv/usbdrvasm.S]:
```suggestion
; 
```

File: usbdrv/usbdrvasm.S [hunk 3]
```diff
@@ -151,7 +153,7 @@ RTMODEL "__rt_version", "3"
 ; unsigned table(unsigned char x)
 ; {
 ; unsigned    value;
-; 
+;
```
Comment (fauxpark) [inline usbdrv/usbdrvasm.S]:
```suggestion
; 
```

File: usbdrv/usbdrvasm.S [hunk 4]
```diff
@@ -117,7 +119,7 @@ RTMODEL "__rt_version", "3"
 #   define polyH    r21
 #   define scratch  r23
 
-#else  /* __IAR_SYSTEMS_ASM__ */ 
+#else  /* __IAR_SYSTEMS_ASM__ */
```
Comment (fauxpark) [inline usbdrv/usbdrvasm.S]:
```suggestion
#else  /* __IAR_SYSTEMS_ASM__ */ 
```

File: usbdrv/usbdrv.c [hunk 5]
```diff
@@ -140,7 +143,7 @@ PROGMEM const char usbDescriptorConfiguration[] = {    /* USB configuration desc
     9,          /* sizeof(usbDescriptorConfiguration): length of descriptor in bytes */
     USBDESCR_CONFIG,    /* descriptor type */
     18 + 7 * USB_CFG_HAVE_INTRIN_ENDPOINT + 7 * USB_CFG_HAVE_INTRIN_ENDPOINT3 +
```
Comment (fauxpark) [inline usbdrv/usbdrv.c]:
```suggestion
    18 + 7 * USB_CFG_HAVE_INTRIN_ENDPOINT + 7 * USB_CFG_HAVE_INTRIN_ENDPOINT3 + 7 * USB_CFG_HAVE_INTRIN_ENDPOINT4 +
```

File: usbdrv/usbdrv.c [hunk 6]
```diff
@@ -140,7 +143,7 @@ PROGMEM const char usbDescriptorConfiguration[] = {    /* USB configuration desc
...
-                (USB_CFG_DESCR_PROPS_HID & 0xff), 0,
+         7 * USB_CFG_HAVE_INTRIN_ENDPOINT4 +(USB_CFG_DESCR_PROPS_HID & 0xff), 0,
```
Comment (fauxpark) [inline usbdrv/usbdrv.c]:
```suggestion
                (USB_CFG_DESCR_PROPS_HID & 0xff), 0,
```

File: usbdrv/usbdrv.c [hunk 7]
```diff
@@ -187,6 +190,14 @@ PROGMEM const char usbDescriptorConfiguration[] = {    /* USB configuration desc
     8, 0,       /* maximum packet size */
     USB_CFG_INTR_POLL_INTERVAL, /* in ms */
 #endif
+#if USB_CFG_HAVE_INTRIN_ENDPOINT4   /* endpoint descriptor for endpoint 3 */
```
Comment (fauxpark) [inline usbdrv/usbdrv.c]:
```suggestion
#if USB_CFG_HAVE_INTRIN_ENDPOINT4   /* endpoint descriptor for endpoint 4 */
```

File: usbdrv/usbdrv.c [hunk 8]
```diff
@@ -187,6 +190,14 @@ PROGMEM const char usbDescriptorConfiguration[] = {    /* USB configuration desc
...
+    7,          /* sizeof(usbDescrEndpoint) */
+    USBDESCR_ENDPOINT,  /* descriptor type = endpoint */
+    (char)(0x80 | USB_CFG_EP4_NUMBER), /* IN endpoint number 3 */
```
Comment (fauxpark) [inline usbdrv/usbdrv.c]:
```suggestion
    (char)(0x80 | USB_CFG_EP4_NUMBER), /* IN endpoint number 4 */
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by yiancar** [comment]
@fauxpark @jackhumbert apparently this needs permissions etc to be reviewed:)

**Review by fauxpark — APPROVED** [review]
Tested on my wonky JJ4x4, can't say anything about larger boards but works like a charm! Nice work @yiancar 👍

**Review by noroadsleft — APPROVED** [review]
`__attribute__ ((weak))` :heavy_check_mark:

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**